va/va.go

```diff
@@ -136,11 +137,48 @@ func (va ValidationAuthorityImpl) getAddr(ctx context.Context, hostname string)
 
 type dialer struct {
 	record core.ValidationRecord
+	stats  metrics.Scope
 }
 
 func (d *dialer) Dial(_, _ string) (net.Conn, error) {
 	realDialer := net.Dialer{Timeout: validationTimeout}
-	return realDialer.Dial("tcp", net.JoinHostPort(d.record.AddressUsed.String(), d.record.Port))
+	conn, err := realDialer.Dial("tcp", net.JoinHostPort(d.record.AddressUsed.String(), d.record.Port))
+
+	// If there was an error connecting to the preferred address and the IPv6First
+	// feature is enabled then attempt another connection if there is a fallback
+	// address available to try
+	if err != nil && features.Enabled(features.IPv6First) {
+		if fallback := fallbackAddress(d.record); fallback != nil {
+			d.record.AddressesTried = append(d.record.AddressesTried, d.record.AddressUsed)
+			d.record.AddressUsed = fallback
+			d.stats.Inc("IPv4Fallback", 1)
+			conn, err = realDialer.Dial("tcp", net.JoinHostPort(fallback.String(), d.record.Port))
+		}
+	}
+
+	return conn, err
+}
+
+// fallbackAddress takes a ValidationRecord and if the `AddressUsed` was IPv6 it
```

Comment:
Catching up this thread with the results of out-of-band discussion: It turns out to be significantly tricky to unit test a network failure for an IPv6 connection that will result in a timeout and not a different error type. We're going to maintain the original behaviour and retry on V4 for any V6 error type, not just a timeout.